### ISSUE TITLE
New context API [Proposal]

### ISSUE THREAD

**Issue by msaraiva** [msaraiva/surface#29]
Similar to `property` and `data`. The goal is to add a new `context` function so we can provide a safer way to use contexts. The new API should:

  1. Classify added variables by parent component, so children cannot accidentally override their values
  1. Allow child components to specify just the variables they need
  1. Merge context variables directly into the assigns
  1. Allow static checking
  1. Consider an added context variable part of the public interface, forcing the author to treat it as such (encourage him to provide documentation for it, etc.)
  1. Export all information regarding contexts for later introspection. This way tools like ElixirSense can provide features like autocomplete, Go-to-definition and documentation preview for various editors.

### Form

```elixir
defmodule Form do
  ...

  @doc "The form data"
  property for, :any, required: true

  @doc "The Form struct created by the <Form/> component"
  context :set, form

  def begin_context(props, context) do
    form = form_for(props.for, ...)
    Map.put(context, :form, form)
  end

  def render(assigns) do
    ~H"""
    {{ @form }}
    ...
    """
  end
  ...
end
```

Pay attention that `form` was automatically merged into the assigns. The `@context.form` is no longer available. Anything available to the user must be in the `assigns` and must have been explicitly declared using the API. Also notice that the `begin_context` now receives the `context` as an extra argument since `props.context` will no longer be available.

### Field

```elixir
defmodule Field do
  ...

  ### Public
  
  @doc "The field name"
  property field, :string, required: true

  @doc "The field name specified in the <Field/> component"
  context :set, field, :string, scope: :children

  ### Private

  context :get, form, from: Form

  def begin_context(props, context) do
    field = String.to_atom(props.field)
    Map.put(context, :field, field)
  end

  def render(assigns) do
    ~H"""
    <div class="field">
      {{ label(@form, @field, class: "label") }}
      ...
    </div>
    """
  end
  ...
end
```

The `scope: :children` option informs that the context variable `field` will only be available to its children, otherwise it would override the existing `field` property. The static checker will not allow overriding existing properties or data and should raise an error. The goal is to force the user to be explicit about the origin of anything present in the assigns.

### TextInput

```elixir
defmodule TextInput do
  ...

  context :get, form, from: Form
  context :get, field, from: Field

  def render(assigns) do
    ~H"""
    {{ text_input(@form, @field, ...) }}
    """
  end
end

```

Feel free to send any comment or suggestion you might have.

@brainlid since you already got your hands dirty with contexts before,  I'd love to hear your thoughts on this one :)

Cheers.

**Comment by msaraiva** [comment]
**Update:**

Since we now have to explicitly declare `context` variables, there's no need to implement the `end_context/1` callback anymore so it will be removed from the API. There will be only an `init_context/1` callback which should return the values to be added. The values will be automatically merged to the context and automatically removed at the end of the scope. Example:

```elixir
context :set, field, :any

def init_context(assigns) do
  {:ok, field: assigns.field}
end
```

**Comment by msaraiva** [comment]
Closed in #44
